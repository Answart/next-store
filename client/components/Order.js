import React from 'react';
import Link from 'next/link';
import PropTypes from 'prop-types';
import { format } from 'date-fns';
import { formatMoney, capWord } from '../lib/utils';
import { StyledOrder } from './styles/OrderStyles';
import { StyledOrderTable } from './styles/TableStyles';
import PriceTag from './PriceTag';


const Order = ({ order }) => (
  <StyledOrder data-test="order">
    <div id="order-header">
      <h2>
        <span>Payment Information</span>
        <span><i>{capWord(order.status)}</i></span>
      </h2>

      <div className="order-payment-details">
        <p>
          <span>Order ID:</span>
          <span>{order.id}</span>
        </p>

        <p>
          <span>Payment ID: </span>
          <span>{order.payment}</span>
        </p>

        <p>
          <span>Order Placed: </span>
          <span>{format(order.createdAt, 'MMMM d, YYYY h:mm a')}</span>
        </p>
      </div>
    </div>

    <StyledOrderTable>
      <thead>
        <tr>
          <th className='column-one'>Product</th>
          <th className='column-two'>Quantity</th>
          <th className='column-three'>Total Price</th>
        </tr>
      </thead>

      <tbody>
        {order.items.map(orderItem => {
          const productId = (!!orderItem.variant && !!orderItem.variant.product)
            ? orderItem.variant.product.id
            : 'unknown';
          return (
            <tr key={orderItem.id}>
              <td className="order-item-details">
                <div className='order-item-image'>
                  <Link href={{
                    pathname: '/buy',
                    query: { id: productId }
                  }}>
                    <a>
                      <img width="100" height="120"
                        src={orderItem.image_url}
                        alt={orderItem.title}
                        title={`Go to: ${orderItem.title}`}
                      />
                    </a>
                  </Link>
                </div>

                <div className='order-item-meta'>
                  <div className="order-item-detail order-item-title">
                    <h4>{orderItem.title}</h4>
                  </div>

                  <div className="order-item-detail">
                    <PriceTag
                      price={orderItem.price}
                      sale={false}
                      salePrice={0}
                    />
                  </div>

                  {orderItem.size && (
                    <div className="order-item-detail">
                      Size: {orderItem.size}
                    </div>
                  )}

                  {orderItem.color && (
                    <div className="order-item-detail">
                      Color: {capWord(orderItem.color)}
                    </div>
                  )}
                </div>
              </td>

              <td className="order-item-quantity">
                {orderItem.quantity}
              </td>

              <td className="order-item-price">
                {formatMoney(orderItem.price * orderItem.quantity)}
              </td>
            </tr>
          )
        })}
      </tbody>

      <tfoot>
        <tr>
          <td></td>
          <td>Subtotal</td>
          <td>{formatMoney(order.subtotal)}</td>
        </tr>

        <tr>
          <td></td>
          <td>Shipping</td>
          <td>{formatMoney(order.shipping)}</td>
        </tr>

        <tr>
          <td></td>
          <td>Sale Tax</td>
          <td>{formatMoney(order.tax)}</td>
        </tr>

        <tr>
          <td></td>
          <td>Grand Total</td>
          <td>{formatMoney(order.total)}</td>
        </tr>
      </tfoot>
    </StyledOrderTable>
  </StyledOrder>
);

Order.propTypes = {
  order: PropTypes.shape({
    id: PropTypes.string.isRequired,
    status: PropTypes.string.isRequired,
    quantity: PropTypes.number.isRequired,
    payment: PropTypes.string.isRequired,
    subtotal: PropTypes.number.isRequired,
    shipping: PropTypes.number.isRequired,
    tax: PropTypes.number.isRequired,
    total: PropTypes.number.isRequired,
    createdAt: PropTypes.string.isRequired,
    items: PropTypes.arrayOf(PropTypes.shape({
      title: PropTypes.string.isRequired,
      quantity: PropTypes.number.isRequired,
      price: PropTypes.number.isRequired,
      size: PropTypes.string,
      color: PropTypes.string,
      image_url: PropTypes.string.isRequired,
      variant: PropTypes.shape({
        product: PropTypes.shape({
          id: PropTypes.string.isRequired,
        })
      })
    })).isRequired,
  }).isRequired,
};


export default Order;
