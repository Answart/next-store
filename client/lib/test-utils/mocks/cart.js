import {
  ADD_TO_CART_MUTATION,
  UPDATE_CARTITEM_MUTATION,
  REMOVE_FROM_CART_MUTATION
} from '../../../graphql'
import { fakeCartItem } from '../utils';
import { mockVariant } from './variant';

const mockCartItem = fakeCartItem();


const addToCartMutationMock = {
  request: { query: ADD_TO_CART_MUTATION, variables: { id: mockVariant.id } },
  result: {
    data: {
      addToCart: {
        __typename: mockCartItem.__typename,
        id: mockCartItem.id,
      },
    },
  },
};

const updateCartItemMutationMock = (quantity) => ({
  request: { query: UPDATE_CARTITEM_MUTATION, variables: { id: mockCartItem.id, quantity } },
  result: {
    data: {
      updateCartItem: {
        __typename: mockCartItem.__typename,
        id: mockCartItem.id,
        quantity,
        variant: { ...mockCartItem.variant }
      },
    },
  },
});

const removeFromCartMutationMock = {
  request: { query: REMOVE_FROM_CART_MUTATION, variables: { id: mockCartItem.id } },
  result: {
    data: {
      removeFromCart: {
        __typename: mockCartItem.__typename,
        id: mockCartItem.id,
      },
    },
  },
};


export {
  mockCartItem,
  addToCartMutationMock,
  updateCartItemMutationMock,
  removeFromCartMutationMock
}
