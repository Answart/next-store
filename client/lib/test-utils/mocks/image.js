import { CREATE_IMAGE_MUTATION } from '../../../graphql'
import { fakeImage } from '../utils';

const mockImage = fakeImage();
const mockImageVariables = {
  cloudinary_id: mockImage.cloudinary_id,
  name: mockImage.name,
  height: mockImage.height,
  width: mockImage.width,
  transformation: mockImage.transformation,
  image_url: mockImage.image_url,
  large_image_url: mockImage.large_image_url
}


const createImageMutationMock = (overrides) => ({
  request: {
    query: CREATE_IMAGE_MUTATION,
    variables: {
      ...mockImageVariables,
      ...overrides,
    },
  },
  result: {
    data: {
      createImage: {
        __typename: 'Image',
        id: mockImage.id,
      },
    },
  },
});

const createImageMutationErrorMock = {
  request: {
    query: CREATE_IMAGE_MUTATION,
    variables: {
      ...mockImageVariables,
      cloudinary_id: '',
      name: '',
    },
  },
  result: {
    errors: [{ message: 'ack!' }],
  }
};


export {
  mockImage, mockImageVariables,
  createImageMutationMock, createImageMutationErrorMock,
}
