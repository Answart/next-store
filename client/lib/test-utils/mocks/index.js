export {
  mockUser, mockUsers,
  mockImage, mockImageVariables,
  mockProduct, mockProducts, mockShopProductsVariables,
  mockVariant,
  mockCartItem,
  mockOrderItem, mockOrder,
} from './typeDefs';
export {
  localUserQueryEmptyCartMock, localUserQueryNoUserMock, localUserQueryCartItemMock,
  userQueryEmptyCartMock, userQueryNoUserMock, userQueryCartItemMock,
  signupMutationMock,
  signinMutationMock,
  signoutMutationMock,
  requestPasswordResetMutationMock, requestPasswordResetMutationErrorMock,
  resetPasswordMutationMock, resetPasswordMutationErrorMock,
  updatePermissionsMutationMock,
  createImageMutationMock, createImageMutationErrorMock,
  productQueryMock, productQueryErrorMock, productQueryNoProductMock, productQueryNoVariantMock,
  shopProductsQueryFilterMock,
  createProductMutationMock, createProductMutationErrorMock,
  updateProductMutationMock, updateProductMutationErrorMock,
  deleteProductMutationMock,
  createProductVariantMutationMock, createProductVariantMutationErrorMock,
  updateProductVariantMutationMock, updateProductVariantMutationErrorMock,
  deleteProductVariantMutationMock,
  addToCartMutationMock,
  updateCartItemMutationMock,
  removeFromCartMutationMock,
} from './resolvers';
