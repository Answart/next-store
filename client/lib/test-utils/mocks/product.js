import {
  PRODUCT_QUERY, SHOP_PRODUCTS_QUERY,
  CREATE_PRODUCT_MUTATION,
  UPDATE_PRODUCT_MUTATION,
  DELETE_PRODUCT_MUTATION,
} from '../../../graphql'
import { mockUser, mockProduct, mockImage, mockVariant, mockShopProductsVariables } from '../utils';


const productQueryMock = {
  request: { query: PRODUCT_QUERY, variables: { id: mockProduct.id } },
  result: {
    data: {
      product: {
        ...mockProduct,
        __typename: mockProduct.__typename,
        image: mockImage,
        user: {
          __typename: mockUser.__typename,
          id: mockUser.id,
          name: mockUser.name,
        },
        variants: [{
          ...mockVariant,
          product: {
            __typename: mockProduct.__typename,
            id: mockProduct.id,
            image: mockImage
          }
        }]
      },
    },
  },
};

const productQueryNoVariantMock = {
  request: { query: PRODUCT_QUERY, variables: { id: mockProduct.id } },
  result: {
    data: {
      product: {
        ...mockProduct,
        __typename: mockProduct.__typename,
        image: mockImage,
        user: {
          __typename: mockUser.__typename,
          id: mockUser.id,
          name: mockUser.name,
        },
        variants: []
      },
    },
  },
};

const createProductMutationMock = {
  request: {
    query: CREATE_PRODUCT_MUTATION,
    variables: {
      department: mockProduct.department,
      title: mockProduct.title,
      description: mockProduct.description,
      category: mockProduct.category,
      brand: mockProduct.brand,
      online: mockProduct.online,
      imageId: mockImage.id,
    },
  },
  result: {
    data: {
      createProduct: {
        __typename: 'Product',
        id: mockProduct.id,
      },
    },
  },
};

const createProductMutationErrorMock = {
  request: {
    query: CREATE_PRODUCT_MUTATION,
    variables: {
      department: 'tops',
      title: '',
      description: '',
      category: '',
      brand: '',
      online: false,
      imageId: mockImage.id,
    },
  },
  result: {
    errors: [{ message: 'ack!' }],
  }
};

const updateProductMutationMock = {
  request: {
    query: UPDATE_PRODUCT_MUTATION,
    variables: {
      id: mockProduct.id,
      department: mockProduct.department,
      title: mockProduct.title,
      description: mockProduct.description,
      category: mockProduct.category,
      brand: mockProduct.brand,
      online: mockProduct.online,
      imageId: mockImage.id,
    },
  },
  result: {
    data: {
      updateProduct: {
        __typename: 'Product',
        id: mockProduct.id,
      },
    },
  },
};

const updateProductMutationErrorMock = {
  request: {
    query: UPDATE_PRODUCT_MUTATION,
    variables: {
      id: mockProduct.id,
      department: 'tops',
      title: '',
      description: '',
      category: '',
      brand: '',
      online: false,
      imageId: mockImage.id,
    },
  },
  result: {
    errors: [{ message: 'ack!' }],
  }
};

const deleteProductMutationMock = {
  request: { query: DELETE_PRODUCT_MUTATION, variables: { id: mockProduct.id } },
  result: {
    data: {
      deleteProduct: {
        __typename: 'Product',
        id: mockProduct.id,
      },
    },
  },
};

const shopProductsQueryNameEmptyMock = {
  request: { query: SHOP_PRODUCTS_QUERY, variables: { name: mockUser.name }},
  result: {
    data: {
      products: [],
    },
  },
};

const shopProductsQueryProductMock = {
  request: { query: SHOP_PRODUCTS_QUERY, variables: { ...mockShopProductsVariables }},
  result: {
    data: {
      products: [{
        ...mockProduct,
        image: mockImage,
        user: {
          __typename: mockUser.__typename,
          id: mockUser.id,
          name: mockUser.name,
        },
        variants: [{
          ...mockVariant,
          product: {
            __typename: mockProduct.__typename,
            id: mockProduct.id,
            image: mockImage,
          }
        }]
      }],
    },
  },
};


export {
  productQueryMock, productQueryNoVariantMock,
  shopProductsQueryNameEmptyMock, shopProductsQueryProductMock,
  createProductMutationMock, createProductMutationErrorMock,
  updateProductMutationMock, updateProductMutationErrorMock,
  deleteProductMutationMock,
}
