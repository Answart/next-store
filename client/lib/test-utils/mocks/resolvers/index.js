export {
  localUserQueryEmptyCartMock, localUserQueryNoUserMock, localUserQueryCartItemMock,
} from './local';
export {
  userQueryEmptyCartMock, userQueryNoUserMock, userQueryCartItemMock,
  signupMutationMock,
  signinMutationMock,
  signoutMutationMock,
  requestPasswordResetMutationMock, requestPasswordResetMutationErrorMock,
  resetPasswordMutationMock, resetPasswordMutationErrorMock,
  updatePermissionsMutationMock,
} from './user';
export {
  createImageMutationMock, createImageMutationErrorMock,
} from './image';
export {
  productQueryMock, productQueryErrorMock, productQueryNoProductMock, productQueryNoVariantMock,
  shopProductsQueryFilterMock,
  createProductMutationMock, createProductMutationErrorMock,
  updateProductMutationMock, updateProductMutationErrorMock,
  deleteProductMutationMock,
} from './product';
export {
  createProductVariantMutationMock, createProductVariantMutationErrorMock,
  updateProductVariantMutationMock, updateProductVariantMutationErrorMock,
  deleteProductVariantMutationMock
} from './variant';
export {
  addToCartMutationMock,
  updateCartItemMutationMock,
  removeFromCartMutationMock
} from './cartItem';
