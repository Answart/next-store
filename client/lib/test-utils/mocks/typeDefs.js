import {
  fakeUser, fakeImage, fakeProduct, fakeVariant, fakeCartItem, fakeOrder, fakeOrderItem,
} from '../utils';


const mockUser = fakeUser();
const mockUser2 = fakeUser();
const mockImage = fakeImage();
const mockProduct = fakeProduct();
const mockVariant = fakeVariant();
const mockCartItem = fakeCartItem();
const mockOrder = fakeOrder();
const mockOrderItem = fakeOrderItem();

mockOrder.items.push(mockOrderItem);

const mockUsers = [
  mockUser,
  {
    ...mockUser2,
    id: '87654321',
    permissions: ['USER'],
  }
];
const mockProducts = [{
  ...mockProduct,
  image: mockImage,
  user: {
    __typename: mockUser.__typename,
    id: mockUser.id,
    name: mockUser.name,
  },
  variants: [{
    ...mockVariant,
    product: {
      __typename: mockProduct.__typename,
      id: mockProduct.id,
      image: mockImage,
    }
  }]
}];

const mockImageVariables = {
  cloudinary_id: mockImage.cloudinary_id,
  name: mockImage.name,
  height: mockImage.height,
  width: mockImage.width,
  transformation: mockImage.transformation,
  image_url: mockImage.image_url,
  large_image_url: mockImage.large_image_url
};

const mockShopProductsVariables = {
  name: mockUser.name,
  orderBy: 'createdAt_DESC',
  skip: 0,
  first: 1
 };


export {
  mockUser, mockUsers,
  mockImage, mockImageVariables,
  mockProduct, mockProducts, mockShopProductsVariables,
  mockVariant,
  mockCartItem,
  mockOrder, mockOrderItem,
};
