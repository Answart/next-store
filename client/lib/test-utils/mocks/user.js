import {
  CURRENT_USER_QUERY,
  SIGNOUT_MUTATION,
  REQUEST_PASSWORD_RESET_MUTATION,
} from '../../../graphql'
import { fakeUser } from '../utils';
import { mockCartItem } from './cart';

const mockUser = fakeUser();


const userQueryEmptyCartMock = {
  request: { query: CURRENT_USER_QUERY },
  result: {
    data: {
      me: {
        ...mockUser,
        cart: [],
      },
    },
  },
};

const userQueryNoUserMock = {
  request: { query: CURRENT_USER_QUERY },
  result: {
    data: { me: null }
  },
};

const userQueryCartItemMock = overrides => ({
  request: { query: CURRENT_USER_QUERY },
  result: {
    data: {
      me: {
        ...mockUser,
        cart: [{
          ...mockCartItem,
          ...overrides
        }],
      },
    },
  },
});

const signoutMutationMock = {
  request: { query: SIGNOUT_MUTATION },
  result: {
    data: {
      signout: {
        __typename: 'Message',
        success: true,
        message: 'Goodbye!'
      },
    },
  },
};

const requestPasswordResetMutationMock = {
  request: { query: REQUEST_PASSWORD_RESET_MUTATION, variables: { email: 'answart@sbcglobal.net' } },
  result: {
    data: {
      requestPasswordReset: {
        __typename: 'Message',
        success: true,
        message: 'Your request has been sent! Check your email.',
      }
    },
  },
};

const requestPasswordResetMutationErrorMock = {
  request: { query: REQUEST_PASSWORD_RESET_MUTATION, variables: { email: 'err@g.c' } },
  result: {
    data: {
      requestPasswordReset: {
        __typename: 'Message',
        success: false,
        message: 'No such user found for email err@g.c',
      }
    },
  },
};


export {
  mockUser,
  userQueryEmptyCartMock, userQueryNoUserMock, userQueryCartItemMock,
  signoutMutationMock,
  requestPasswordResetMutationMock, requestPasswordResetMutationErrorMock
};