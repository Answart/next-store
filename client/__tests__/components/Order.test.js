import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import wait from 'waait';
import { MockedProvider } from 'react-apollo/test-utils';
import Order from '../../components/Order';
import { mockOrder } from '../../lib/test-utils/mocks';


describe('<Order />', () => {
  let wrapper;
  beforeAll(() => {
    wrapper = mount(
      <MockedProvider mocks={[]}>
        <Order order={mockOrder} />
      </MockedProvider>
    );
  });
  afterAll(() => {
    wrapper.unmount();
    jest.clearAllMocks();
  });

  it('matches the snap shot', async () => {
    await wait();
    wrapper.update();
    expect(toJSON(wrapper.find('form[data-test="order"]'))).toMatchSnapshot();
  });

  it('renders properly', async () => {
    const header = wrapper.find('#order-header')
    expect(header.find('h2').length).toBe(1);
    expect(header.find('h2').find('span').length).toBe(2);
    expect(header.find('.order-payment-details').length).toBe(1);
    expect(header.find('.order-payment-details').find('p').length).toBe(3);
    const table = wrapper.find('table');
    expect(table.length).toBe(1);
    expect(table.find('thead').length).toBe(1);
    expect(table.find('thead').find('th').length).toBe(3);
    expect(table.find('tbody').length).toBe(1);
    expect(table.find('tbody').find('tr').length).toBe(1);
    expect(table.find('tbody').find('td').length).toBe(3);
    expect(table.find('tbody').find('Link').props().href).toMatchObject({
      pathname: "/buy",
      query: { id: mockOrder.items[0].variant.product.id }
    });
    expect(table.find('tfoot').length).toBe(1);
    expect(table.find('tfoot').find('tr').length).toBe(4);
  });
});
