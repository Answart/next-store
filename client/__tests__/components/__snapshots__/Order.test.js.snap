// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Order /> matches the snap shot 1`] = `null`;
